docs/src/main/asciidoc/security-openid-connect-client-registration.adoc

```diff
@@ -0,0 +1,332 @@
+
+The above configuration will lead to two new client registrations created in your OIDC provider.
+
+You or may not need to acquire an initial registration access token. If you don't, then you will have to enable one or more client registration policies in your OIDC provider's dashboard. For example, see https://www.keycloak.org/docs/latest/securing_apps/#_client_registration_policies[Keycloak client registration policies].   
+
+The next step is to inject either `quarkus.oidc.client.registration.OidcClientRegistration` if only a single default client registration is done, or `quarkus.oidc.client.registration.OidcClientRegistrations` if more than one registration is configured, and use metadata of the clients registered with these configurations.
+
+For example:
+
+[source,java]
+----
+package io.quarkus.it.keycloak;
+
+import java.net.URI;
+import java.util.List;
+import java.util.Optional;
+
+import jakarta.enterprise.event.Observes;
+import jakarta.inject.Inject;
+import jakarta.inject.Singleton;
+
+import org.eclipse.microprofile.config.inject.ConfigProperty;
+
+import io.quarkus.oidc.OidcRequestContext;
+import io.quarkus.oidc.OidcTenantConfig;
+import io.quarkus.oidc.OidcTenantConfig.ApplicationType;
+import io.quarkus.oidc.TenantConfigResolver;
+import io.quarkus.oidc.client.registration.ClientMetadata;
+import io.quarkus.oidc.client.registration.OidcClientRegistration;
+import io.quarkus.oidc.client.registration.OidcClientRegistrationConfig;
+import io.quarkus.oidc.client.registration.OidcClientRegistrations;
+import io.quarkus.oidc.client.registration.RegisteredClient;
+import io.smallrye.mutiny.Uni;
+import io.vertx.ext.web.RoutingContext;
+
+@Singleton
+public class CustomTenantConfigResolver implements TenantConfigResolver {
+
+    @Inject
+    OidcClientRegistration clientReg;
+
+    @Inject
+    OidcClientRegistrations clientRegs;
+
+    @Override
+    public Uni<OidcTenantConfig> resolve(RoutingContext routingContext,
+            OidcRequestContext<OidcTenantConfig> requestContext) {
+            
+        if (routingContext.request().path().endsWith("/protected")) {
+            // Use the registered client created from the default OIDC client registration 
+            
+            return Uni.createFrom().item(createTenantConfig("registered-client", clientReg.registeredClient()));
+            
+        } else if (routingContext.request().path().endsWith("/protected/tenant")) {
+            // Use the registered client created from the named 'tenant-client' OIDC client registration
+            
+            return Uni.createFrom().item(createTenantConfig("registered-client-tenant",
+                clientRegs.getClientRegistration("tenant-client").registeredClient()));
+        }
+        return null;
+    }
+
+    // Convert metadata of registered clients to OidcTenantConfig
+    private OidcTenantConfig createTenantConfig(String tenantId, RegisteredClient client) {
+        ClientMetadata metadata = client.getMetadata();
```

Comment:
What if we make it possible to create a `OidcTenantConfig` from the metadata as a built-in capability in `quarkus-oidc`? The code here should repeat a lot when using custom tenants based on registered clients.

Perhaps, even the the `application_type` claim from the client metadata can be used to infer the `ApplicationType` for the quarkus application.

Reply:
+1, in fact the initial draft PR content did try to do it, but then I thought it was getting too complex for a single PR.

Note, right now, the new `oidc-client-registration` extension is totally independent of `quarkus-oidc`, users can build a system application which only facilitates the client registration, makes the new client registration details visible to `quarkus-oidc` aware services via some medium like secret engines, which is what we touched on with @calvernaz, etc.

But yeah, I also had another extension prototyped, `oidc-client-registration-config-source` where, at the startup, the client is registered and `quarkus-oidc` has all its properties like client id, client secret, auto-resolved from those metadata. It was in the early draft PR, but then I thought, let's do it afterwards, once the core client registration API settles, and then we can make more integral to `quarkus-oidc`